### DISCUSSION TITLE
the Flutter web version of project-m36-websocket-client

### DISCUSSION THREAD

**Discussion by YuMingLiao** [agentm/project-m36#329] [Show and tell]
In case anyone is interested,
[here](https://github.com/YuMingLiao/project_m36_websocket_client) is a Flutter web version of project-m36-websocket-client. 

It's also an attempt to bring ADTs to the Dart land.

**Comment by agentm** [comment]
Very cool! Could you share some screenshots?

**Reply by YuMingLiao** [reply]
Added.

**Reply by YuMingLiao** [reply]
hi @agentm 
Would you happen to know that if it's possible to use project-m36 in android as a local database?
I would like to use project-m36 as a flutter app's local database.

**Reply by YuMingLiao** [reply]
Okay, I think first I should try initializing a project-m36 executable by flutter's [process](https://api.flutter.dev/flutter/dart-io/Process-class.html) and see if they can communicate.

**Reply by agentm** [reply]
Oh, I have never attempted to build Project:M36 on Android. I would be willing to collaborate on this and would definitely accept patches to build on new platforms.

**Reply by YuMingLiao** [reply]
Glad to find someone interested!

I have a remote websocket-client Flutter app now.
My goal is to install a Flutter app shipping/bundle with a local project-m36 database for itself.
Now, I don't know how to make Flutter, Haskell and Android work together.

What I know:

- [sqflite](https://pub.dev/packages/sqflite) doesn't ship/bundle a sqlite. So no code for borrowing there.
- [haskell on android](https://wiki.haskell.org/Android)
- Flutter has gRPC. ProjectM36 uses currier-rpc. And Haskell has a gRPC library.
- Executables may need to be recompiled by NDK for specific libraries on android.
- [flutter_background_service](https://pub.dev/packages/flutter_background_service)
- [android-haskell-activity](https://github.com/neurocyte/android-haskell-activity)

What I don't know:

- how to bring  project-m36 into an Android environment.
- how to make Flutter app communicates with project-m36. Can Android have local websocket communication?
  - [Writing custom platform-specific code](https://docs.flutter.dev/development/platform-integration/platform-channels) 
- how to make a Flutter app to ship/bundle with project-m36
  - [how-to-bundle-native-macos-executable-with-a-flutter-app](https://stackoverflow.com/questions/60942772/how-to-bundle-native-macos-executable-with-a-flutter-app)
  - [hosting-an-executable-within-a-flutter-application](https://stackoverflow.com/questions/62056764/hosting-an-executable-within-a-flutter-application)







And I've tried:
```
126|RMX2020:/data/local/tmp $ chmod 777 tutd
RMX2020:/data/local/tmp $ ./tutd
sh: ./tutd: not executable: 64-bit ELF file
````
[android-not-executable-64-bit-elf-file](https://stackoverflow.com/questions/41644382/android-not-executable-64-bit-elf-file)

**Reply by YuMingLiao** [reply]
I expect a lot of bumps on the road of re-compilation on android. I will accept a remote server to continue to have fun making my app.

**Reply by YuMingLiao** [reply]
Related References for the issue: 
[A Haskell Cross Compiler for Android](https://medium.com/@zw3rk/a-haskell-cross-compiler-for-android-8e297cb74e8a?p=7e38c00c2914)
[docker-game-build-env](https://github.com/sseefried/docker-game-build-env)
[Keera Studios: Haskell on Android and iOS](https://keera.co.uk/2017/06/01/haskell-android-ios/)
